drivers/cache/cache_andes.c

```diff
@@ -0,0 +1,737 @@
+	} else {
+		cache_cfg.l2c_cmd_offset = 0x1000;
+		cache_cfg.l2c_status_offset = 0x1000;
+		cache_cfg.l2c_status_shift = 0;
+	}
+
+	cache_cfg.l2c_version = (sys_read32(L2C_CONFIG) >> 24) & BIT_MASK(8);
+
+	/* Initializing L2 cache instruction, data prefetch depth */
+	l2c_ctrl = sys_read32(L2C_CTRL);
+	l2c_ctrl |= ((3 << 5) | (3 << 3));
+
+	/* Writeback and invalidate all I/D-Cache before setting L2C */
+	__asm__ volatile ("fence.i");
+	sys_write32(l2c_ctrl, L2C_CTRL);
+
+#ifdef CONFIG_SMP
+	if (cache_cfg.l2c_size) {
+		uint32_t l2c_ctrl = sys_read32(L2C_CTRL);
+
+		if (!(l2c_ctrl & BIT(0))) {
+			l2c_ctrl |= BIT(0);
+			sys_write32(l2c_ctrl, L2C_CTRL);
+		}
+	}
+#endif /* CONFIG_SMP */
+
+	return 0;
+}
+
+#endif /* L2C_EXISTS */
+
+static ALWAYS_INLINE void nds_l1i_cache_all(int op)
+{
+	unsigned long sets, ways, end;
+
+	if (op == K_CACHE_INVD) {
+		sets = 0x40 << (csr_read(NDS_MICM_CFG) & BIT_MASK(3));
+		ways = ((csr_read(NDS_MICM_CFG) >> 3) & BIT_MASK(3)) + 1;
+		end = ways * sets * cache_cfg.instr_line_size;
+
+		for (int i = 0; i < end; i += cache_cfg.instr_line_size) {
+			csr_write(NDS_MCCTLBEGINADDR, i);
+			csr_write(NDS_MCCTLCOMMAND, CCTL_L1I_IX_INVAL);
+		}
+	}
+}
+
+static ALWAYS_INLINE void nds_l1d_cache_all(int op)
+{
+	switch (op) {
+	case K_CACHE_WB:
+		csr_write(NDS_MCCTLCOMMAND, CCTL_L1D_WB_ALL);
+		break;
+	case K_CACHE_INVD:
+		csr_write(NDS_MCCTLCOMMAND, CCTL_L1D_INVAL_ALL);
+		break;
+	case K_CACHE_WB_INVD:
+		csr_write(NDS_MCCTLCOMMAND, CCTL_L1D_WBINVAL_ALL);
+		break;
+	default:
+		return;
+	}
+}
+
+static ALWAYS_INLINE void nds_l1i_cache_range(void *addr, size_t size, int op)
+{
+	unsigned long last_byte, align_addr, cmd;
+
+	if (op == K_CACHE_INVD) {
+		cmd = CCTL_L1I_VA_INVAL;
+		last_byte = (unsigned long)addr + size - 1;
+		align_addr = ROUND_DOWN(addr, cache_cfg.instr_line_size);
+
+		while (align_addr <= last_byte) {
+			csr_write(NDS_MCCTLBEGINADDR, align_addr);
+			csr_write(NDS_MCCTLCOMMAND, cmd);
+			align_addr += cache_cfg.instr_line_size;
+		}
+	}
+}
+
+static ALWAYS_INLINE void nds_l1d_cache_range(void *addr, size_t size, int op)
+{
+	unsigned long last_byte, align_addr, cmd;
+
+	switch (op) {
+	case K_CACHE_WB:
+		cmd = CCTL_L1D_VA_WB;
+		break;
+	case K_CACHE_INVD:
+		cmd = CCTL_L1D_VA_INVAL;
+		break;
+	case K_CACHE_WB_INVD:
+		cmd = CCTL_L1D_VA_WBINVAL;
+		break;
+	default:
+		return;
+	}
+
+	last_byte = (unsigned long)addr + size - 1;
+	align_addr = ROUND_DOWN(addr, cache_cfg.data_line_size);
+
+	while (align_addr <= last_byte) {
+		csr_write(NDS_MCCTLBEGINADDR, align_addr);
+		csr_write(NDS_MCCTLCOMMAND, cmd);
+		align_addr += cache_cfg.data_line_size;
+	}
+}
+
+void cache_data_enable(void)
+{
+#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)
+	return;
+#endif
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
+
+#ifdef L2C_EXISTS
+	if (cache_cfg.l2c_size) {
+		uint32_t l2c_ctrl = sys_read32(L2C_CTRL);
+
+		if (!(l2c_ctrl & BIT(0))) {
+			l2c_ctrl |= BIT(0);
```

Comment:
we have macros for these, if you like:
```suggestion
			WRITE_BIT(l2c_ctrl, 0, true);
```

Reply:
Thanks! All of these are useful!
We've considered moving L2 related logics to another file, but are not sure whether it is acceptable regarding having multiple C files for a single driver.